File: fn/src/main/java/org/sqlbuilder/fn/joins/Frame_SemType.java
package org.sqlbuilder.fn.joins;

import org.sqlbuilder.common.Insertable;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Set;

public class Frame_SemType extends Pair<Integer, Integer> implements Insertable<Frame_SemType>
{
	public static final Set<Frame_SemType> SET = new HashSet<>();

	// C O N S T R U C T O R

	public static Frame_SemType make(final int frameid, final int semtypeid)
	{
		var ft = new Frame_SemType(frameid, semtypeid);
		SET.add(ft);
		return ft;
	}

	private Frame_SemType(final int frameid, final int semtypeid)
	{
		super(frameid, semtypeid);
	}

	// O R D E R

	public static final Comparator<Frame_SemType> COMPARATOR = Comparator.comparing(Frame_SemType::getFirst).thenComparing(Frame_SemType::getSecond);

	// I N S E R T

	@Override
	public String dataRow()
	{
		return String.format("%d,%d", first, second);
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[FR-SEM frameid=%s semtypeid=%s]", first, second);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/joins/Frame_SemType.kt
package org.sqlbuilder.fn.joins

import org.sqlbuilder.common.Insertable
import java.util.*

data class Frame_SemType(
    val frameid: Int,
    val semtypeid: Int,
) : Insertable {

    // I N S E R T

    override fun dataRow(): String {
        return "$frameid,$semtypeid"
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as Frame_SemType
        return frameid == that.frameid && semtypeid == that.semtypeid
    }

    override fun hashCode(): Int {
        return Objects.hash(frameid, semtypeid)
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[FR-SEM frameid=$frameid semtypeid=$semtypeid]"
    }

    companion object {

        @JvmField
        val COMPARATOR: Comparator<Frame_SemType> = Comparator
            .comparing<Frame_SemType, Int> { it.frameid }
            .thenComparing<Int> { it.semtypeid }

        @JvmField
        val SET = HashSet<Frame_SemType>()

        @JvmStatic
        fun make(frameid: Int, semtypeid: Int): Frame_SemType {
            val ft = Frame_SemType(frameid, semtypeid)
            SET.add(ft)
            return ft
        }
    }
}
